### PR TITLE
iOS promotion test

### PR DESCRIPTION
**Pull request by rayborn**
Population:
- en-* & geo:  CA, NZ, Australia (negotiable.)
- see exactly 1 of 3 possible messages.

### CODE REVIEW

File: src/recipes/ios-promo/index.js [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    name:     BASE_NAME + "_Other",
+    prompt:   "Bring your bookmarks & passwords with you. Firefox is now on iOS & Android.",
+    thankyou: thankyou,
+    url:      url,
+    button:   button
+  }
+];
+
+//Gets a random branch (even distribution)
+let getbranch = function(branches) {
```
Comment (rayborn) [inline src/recipes/ios-promo/index.js]: Needs a closer look

File: src/recipes/ios-promo/index.js [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+/*eData = {
+  data: {
+    flows: {},
+    lastRun: number
+  }
+},
+*/
+
+// Populates a global state by looping over the known discrete states
+var setupGlobalState = function (key, storage) {
```
Comment (rayborn) [inline src/recipes/ios-promo/index.js]: Needs a closer look

**Comment by gregglind** [comment]
Maybe just doing locale=en-GB is safer?  Geo is... dicey

**Comment by gregglind** [comment]
@rayborn,

Lots of news.  

1.  I *think* that simple messages can phone home.
2.  If so, we need to only do so *sometimes* or we will MELT input.
3.  I am going to augment the `common/heartbeat/ui.js` to expose this for the 'real deploy'
4.  This probably means you should copy from the full heartbeat recipe, and only phonehome *sometimes* (say 0.01%)?
5.  Ps, I hate that we resolve sampling at the client (in general!).  I will ask @willkg about filtering at input, but I think it's on us.

**Comment by gregglind** [comment]
@rayborn, simpler solution.

 if you just want to 'ping home', I think the right thing to do is in your run (after elibility), crib these sorts of lines (from https://github.com/mozilla/self-repair-server/blob/master/src/recipes/heartbeat-by-user-first-impression/index.js#L177-L189 and nearby)

``` 
  // make and setup flow
  let flow = new Flow(local);  // create and update
  flow.began();
  maybePhonehome(flow);
```

If you do this for some small (known) % of users, right before you pop the message, then you have the denominator for your study :)

New algorithm:

1.  decide to show (shouldRun)
2.  wait () some random amount of time (if desired, I wouldn't actually bother, complicates things to no real benefit).
3.  GO!  (run)
     a.  phoneHome()  (for a small % of users!  to get denominator, from input)
     b.  pop the message
     c.  user will click (use GA at the site for numerator)



You actually don't need

**Comment by gregglind** [comment]
"You don't actuaally need".... what?  I Don't know!

Another strategy is to get the aliveness from telemetry each day. 
- That is doable (medium easily) from the executive stream, but has *geo* but not locale.
- doable from `get_pings` but slower.  (but has country)

**Comment by gregglind** [comment]
(So to make this a simple as possible)

```
- config
  - sample and rest days.  also exports marketing messages array?
  - index
    - lstore: marketing messages
    - shouldRun() :  
      - checks the lstore (marketing) for restdays
      - checks for 'already seen this particular message' ?   # gross?
    - run()
      - NO phone home (use telemetry instead)
      - pop the message
       - record 'seen' in lstore
```

I think this is the fastest path to landing and launch.

File: src/recipes/ios-promo/config.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  {
+    name:     "bookmarks",
+    prompt:   "Bring your bookmarks & passwords with you. Firefox is now on iOS & Android.",
+    thankyou: thankyou,
+    url:      url,
+    button:   button
+  }
+];
+
+let branch = getbranch(branches);
```
Comment (gregglind) [inline src/recipes/ios-promo/config.js]: I claim this should happen at the recipe.  Doing it here makes it hard to test this file.

File: test/test-built/test-built-exports.js [hunk 4]
```diff
@@ -41,9 +41,10 @@ describe("built file exports", function () {
       expect(heartbeat[k]).to.be.a(mytype);
     }
   });
+  let numRecipes = 3;
```
Comment (gregglind) [inline test/test-built/test-built-exports.js]: I like the instinct here.  I am not sure that I like the solution.  

Claim:  adding a recipe should be 'hard' and a bit of a hassle. I have no good reason for rejecting this, other than searching for the error message string will fail here.

File: src/recipes/ios-promo/config.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    key: 'marketing',
+    version: 1,
+    delay: 60 * 1000 * (1 + 4*Math.random()), // Delay the start by 1-5 minutes
+    //phonehomepct: .01*percent
+  },
+  channels: {
+    all: {
+      sample: 1 * percent, // will be 1.0 when we launch
+      restdays: 1000,
+      locales: ["en-CA"], // will be en-* locales when we launch
```
Comment (gregglind) [inline src/recipes/ios-promo/config.js]: 07:32 < bwinton> gregglind: Well, I don't think we ship Firefox in en-CA…
07:32 < bwinton> So most of us just suffer through eb-GB with its odd spelling of "tyre" or en-US with its odd spelling of
                 "color"…

File: src/recipes/ios-promo/index.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+      fullUrl,
+      null, // learn more text, TODO: should this be "What's this?" or something?
+      null, // learn more link
+      {
+        engagementButtonLabel: BRANCH.button
+      }
+    )},
+    delay
+  );
+  return Promise.resolve(local.flow_id);
```
Comment (gregglind) [inline src/recipes/ios-promo/index.js]: So, this.

`run` is always a promise.  I need to think a bit on whether you should do the promise here, or resolve it inside the timeout.  I did not think this case through.

Comment (rayborn) [inline src/recipes/ios-promo/index.js]: This is straight out of PBM BTW.  Not that that guarantees that I did the right thing :)

File: test/test-built/test-built-exports.js (hunk 4 above)
Comment (rayborn) [inline test/test-built/test-built-exports.js]: Fair, I was trying to make it easy, but I get the couterargument

File: src/recipes/ios-promo/config.js (hunk 3 above)
Comment (rayborn) [inline src/recipes/ios-promo/config.js]: That's fair.  I wanted all branch logic in the same place, but you're right about testing.

**Comment by gregglind** [comment]
closed in favor of #180

**Pull request closed without merging**